feat: Pause media upload when internet connection unavailable

dcalhoun opened the pull request:
## Related 

* https://github.com/WordPress/gutenberg/pull/56937
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6491

## Description

Pause media upload when internet connection is unavailable.

## Testing Instructions
See https://github.com/WordPress/gutenberg/pull/56937.

## Regression Notes
1. Potential unintended areas of impact
    Disrupting successful media uploads, breaking other media block types. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
    Uploaded media to various block types. 
3. What automated tests I added (or what prevented me from doing so)
    No, opted to test the JavaScript logic instead. 

## PR Submission Checklist

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## UI Changes Testing Checklist

- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] Talkback.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

dcalhoun committed:
feat: Failed uploads while offline communicate paused state

This is an exploration. The approach taken, namely using the error type
string, is not fully tested.

dcalhoun committed:
build: Update Gutenberg ref

dcalhoun committed:
build: Fix incorrect Gutenberg PR reference

dcalhoun committed:
build: Fix incorrect Gutenberg PR reference

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/EditorMedia.kt:
```diff
@@ -295,7 +295,7 @@ class EditorMedia @Inject constructor(
         }
         analyticsTrackerWrapper.track(EDITOR_UPLOAD_MEDIA_FAILED, properties)
-        listener.onMediaUploadFailed(media.id.toString())
+        listener.onMediaUploadFailed(media.id.toString(), error.type.name)
```
Passing solely the `error.type.name` string feels a bit messy, but when I attempt to pass `error` and I am unable to leverage the relevant types from the FluxC package, presumably because the editor files are in a separate `lib` from the main app. 

Is it possible to access the `MediaError` type from the FluxC package?

derekblank replied:
> Is it possible to access the MediaError type from the FluxC package?

Yes. In other areas, `MediaErrorType` is imported directly from the `MediaStore`. However, we already have a precedent of sending `error.type.name` from `MediaError` to the analytics in this function, so I'm not certain if there's an advantage of importing `MediaErrorType` directly here. It seems fine to pass `error.type.name` to `onMediaUploadFailed` here, but certainly open to using `MediaErrorType` if you see an advantage.

https://github.com/wordpress-mobile/WordPress-Android/blob/234db297728a506ee8e9c318d42c9cd9e7e941b3/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L107-L108

derekblank replied:
As noted in https://github.com/wordpress-mobile/WordPress-Android/pull/19878#discussion_r1441396013, we'd avoid the need to pass `error.type.name` entirely in https://github.com/wordpress-mobile/WordPress-Android/pull/19884.

dcalhoun commented on libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -1483,8 +1487,15 @@ public void onMediaUploadProgress(final String localMediaId, final float progres
-        getGutenbergContainerFragment().mediaFileUploadFailed(Integer.valueOf(localMediaId));
+    public void onMediaUploadFailed(final String localMediaId, String errorType) {
+        switch (errorType) {
+            case "CONNECTION_ERROR":
```
Using `CONNECTION_ERROR` is based off it being the only error type I observed while disabling my device's network connectivity with airplane mode. I'm uncertain if there are other error types that might be thrown.

derekblank replied:
> I'm uncertain if there are other error types that might be thrown.

In my testing, I did not observe any other error type beyond `CONNECTION_ERROR`. I did note the existence of another similar generic MediaErrorType called [`NO_CONNECTION`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java#L498).

Perhaps we should check for both:

```
switch (errorType) {
    case "CONNECTION_ERROR":
    case "NO_CONNECTION":
        getGutenbergContainerFragment().mediaFileUploadPaused(Integer.valueOf(localMediaId));
        break;
    default:
        getGutenbergContainerFragment().mediaFileUploadFailed(Integer.valueOf(localMediaId));
        break;
}
```

@ParaskP7 As the author of `NO_CONNECTION`, do you have any perspective on when `NO_CONNECTION` may be thrown instead of `CONNECTION_ERROR`? For some quick context, we are attempting to add a new  media upload state called "paused" when the device is offline (to invoke `mediaFileUploadPaused`, instead of marking the upload failed with `mediaFileUploadFailed`).

derekblank replied:
Unless I misunderstand how failed media uploads work, we may not need to wait for an error to be thrown, and could instead listen to network status instead. I have outlined a potential example in https://github.com/wordpress-mobile/WordPress-Android/pull/19884 that augments this PR. If we chose to merge that change into this PR, we would avoid this section of code entirely, and therefore the differences between `NO_CONNECTION` and `CONNECTION_ERROR` (or error types at all) would not be relevant.

I would still be interested in learning the circumstances of when `NO_CONNECTION` can be thrown, and also learning if there is a need to wait for an explicit connection error to be thrown that I'm not considering (instead of just listening to the connection status).

derekblank committed:
Add discrete onMediaUploadPaused handler to media uploads when offline

derekblank reviewed (changes requested):
@dcalhoun Thanks for adding this change! I have tested and it is working quite well to update the Image block visuals when offline, as intended. 👍 🙇 

I've proposed taking it a step further in https://github.com/wordpress-mobile/WordPress-Android/pull/19884 to add a truly discrete `onMediaUploadPaused` handler. That PR augments this one, and is in the spirit of what I was initially trying to accomplish in https://github.com/wordpress-mobile/WordPress-Android/pull/19867.

Let me know your thoughts on using a more discrete paused handler -- if it's valuable, feel free to have full agency to merge or modify that PR here in the interest of optimizing time, or just take any parts that are useful. If not, I feel this PR itself is enough to approve and merge after the next release cut on Jan 8.

dcalhoun committed:
refactor: Fix parameter name typo

dcalhoun committed:
refactor: Annotate override of `onMediaUploadPaused` method

Ensure the implementation adheres to the relevant interface.

dcalhoun committed:
refactor: Remove unused `onMediaUploadFailed` error type parameter

With the addition of a discrete paused method, we no longer utilize this
parameter in the failed method.

dcalhoun committed:
feat: Aztec `onMediaUploadPaused` passes onto `onMediaUploadFailed`

We do not plan to add paused media support to the Aztec editor.
Therefore, it likely makes more sense for the `onMediaUploadPaused`
implementation to merely invoke the existing `onMediaUploadFailed`
method.

dcalhoun committed:
feat: Only consider failed media uploads to be paused

Without checking for an error, there may be a possibility that
successfully uploaded media is marked as "paused."

dcalhoun committed:
feat: Track failed media upload event when pausing media uploads

Even though we communicate a nuanced paused state to the user, we likely
need to continue tracking this event as a failure, or consider adding a
discrete pause analytic event.

dcalhoun committed:
feat: When lacking connectivity, pause completed upload w/o remote URLs

Based upon existing code, we should consider completed uploads lacking
remote URLs to be failures. Therefore, we should apply the new pause
state to this context.

dcalhoun committed:
Merge pull request #19884 from wordpress-mobile/feat/update-image-block-upload-visuals__discrete-paused-state

Add discrete onMediaUploadPaused handler to media uploads when offline

dcalhoun committed:
docs: Add release note

dcalhoun committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into feat/update-image-block-upload-visuals

dcalhoun committed:
build: Update Gutenberg ref

derekblank commented on WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/EditorMedia.kt:
```diff
@@ -295,7 +295,19 @@ class EditorMedia @Inject constructor(
+                    it["error_type"] = error.type.name
+                }
+        }
+        analyticsTrackerWrapper.track(EDITOR_UPLOAD_MEDIA_FAILED, properties)
```
This will still send an `EDITOR_UPLOAD_MEDIA_FAILED` event to analytics when the upload is paused. I believe we'd want to remove these lines, but wanted to double check.

dcalhoun replied:
My thought is that we should still track these as errors (or, at least, track something). I referenced this in https://github.com/wordpress-mobile/WordPress-Android/pull/19884#discussion_r1441786432. I imagine knowing the rate of this event is still useful for analysis. 

Will you expand on why you think we should not?

derekblank replied:
> Will you expand on why you think we should not?

My reasoning was that it might incorrectly affect the metrics outlined in the pitch, which were aiming to reduce instances of `media_service_upload_failed` and `media_service_upload_response_error`. I haven't yet checked to see if `media_service_upload_failed` results are actually affected by `analyticsTrackerWrapper.track(EDITOR_UPLOAD_MEDIA_FAILED, properties)`, but it seems like it might be. 

When the media upload is paused, I think we would expect it to retry and continue when the connection is resumed. If the upload then reached a legitimate failure, it would be captured by `onMediaUploadFailed`, which would fire the analytics failed event. Does that sound correct?

Either way, I don't think it should be a blocker for this PR. I've created an [issue on the project board](https://github.com/orgs/wordpress-mobile/projects/213/views/1?layout=board&pane=issue&itemId=49075278) to discuss adding a paused analytics event.

dcalhoun replied:
Thank you for elaborating. Your rationale makes complete sense. I had not considered the impact on the planned project metrics. 

I agree that we should follow up to remove or change this analytic event. We should also verify that the iOS implementation is correctly set up as well. I'll update your newly created issue to note this.

derekblank reviewed (approved):
Added https://github.com/wordpress-mobile/WordPress-Android/pull/19878#discussion_r1442410956 about potentially removing the analytics event when the upload is paused. Otherwise, LGTM. 🚀

derekblank committed:
Merge branch 'trunk' into feat/update-image-block-upload-visuals

dcalhoun committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into feat/update-image-block-upload-visuals

dcalhoun committed:
build: Update Gutenberg ref
